README.md
# Rendertron [![Build status](https://travis-ci.org/GoogleChrome/rendertron.svg?branch=master)](https://travis-ci.org/GoogleChrome/rendertron) [![NPM rendertron package](https://img.shields.io/npm/v/rendertron.svg)](https://npmjs.org/package/rendertron)

> Rendertron is a headless Chrome rendering solution designed to render & serialise web pages on the fly.

#### :hammer: Built with [Puppeteer](https://github.com/GoogleChrome/puppeteer)

#### :cloud: Easy deployment to Google Cloud

#### :mag: Improves SEO

fork for ammobin
uses: node 8 base image + exposes port 3000 + updated docker file to build on my machine (docker -v
Docker version 17.12.0-ce)

```
docker pull ammobindotca/rendertron
docker run -p 3000:3000 ammobindotca/rendertron
```

https://hub.docker.com/r/ammobindotca/rendertron/

Rendertron is designed to enable your Progressive Web App (PWA) to serve the correct
content to any bot that doesn't render or execute JavaScript. Rendertron runs as a
standalone HTTP server. Rendertron renders requested pages using Headless Chrome,
[auto-detecting](#auto-detecting-loading-function) when your PWA has completed loading
and serializes the response back to the original request. To use Rendertron, your application
configures [middleware](#middleware) to determine whether to proxy a request to Rendertron.
Rendertron is compatible with all client side technologies, including [web components](#web-components).

**Demo endpoint**

A demo Rendertron service is available at https://render-tron.appspot.com/. It is not designed
to be used as a production endpoint. You can use it, but there are no uptime guarantees.

## Contents

-   [Middleware](#middleware)
-   [API](#api)
    -   [Render](#render)
    -   [Screenshot](#screenshot)
-   [FAQ](#faq)
    -   [Query parameters](#query-parameters)
    -   [Auto detecting loading function](#auto-detecting-loading-function)
    -   [Rendering budget timeout](#rendering-budget-timeout)
    -   [Web components](#web-components)
    -   [Status codes](#status-codes)
-   [Installing & deploying](#installing--deploying)
    -   [Building](#building)
    -   [Running locally](#running-locally)
    -   [Deploying to Google Cloud Platform](#deploying-to-google-cloud-platform)
    -   [Deploying using Docker](#deploying-using-docker)
    -   [Config](#config)
    -   [Troubleshooting](#troubleshooting)

## Middleware

Once you have the service up and running, you'll need to implement the differential serving
layer. This checks the user agent to determine whether prerendering is required.

This is a list of middleware available to use with the Rendertron service:

-   [Express.js middleware](/middleware)
-   [Firebase functions](https://github.com/justinribeiro/pwa-firebase-functions-botrender) (Community maintained)
-   [ASP.net core middleware](https://github.com/galamai/AspNetCore.Rendertron) (Community maintained)

Rendertron is also compatible with [prerender.io middleware](https://prerender.io/documentation/install-middleware).
Note: the user agent lists differ there.

## API

### Render

```
GET /render/<url>
```

The `render` endpoint will render your page and serialize your page. Options are
specified as query parameters:

-   `mobile` defaults to `false`. Enable by passing `?mobile` to request the
    mobile version of your site.

### Screenshot

```
GET /screenshot/<url>
POST /screenshot/<url>
```

The `screenshot` endpoint can be used to verify that your page is rendering
correctly.

Both endpoints support the following query parameters:

-   `width` defaults to `1000` - specifies viewport width.
-   `height` defaults to `1000` - specifies viewport height.
-   `mobile` defaults to `false`. Enable by passing `?mobile` to request the
    mobile version of your site.

Additional options are available as a JSON string in the `POST` body. See
[Puppeteer documentation](https://github.com/GoogleChrome/puppeteer/blob/v1.6.0/docs/api.md#pagescreenshotoptions)
for available options. You cannot specify the `type` (defaults to `jpeg`) and
`encoding` (defaults to `binary`) parameters.

## FAQ

### Query parameters

When setting query parameters as part of your URL, ensure they are encoded correctly. In JS,
this would be `encodeURIComponent(myURLWithParams)`. For example to specify `page=home`:

```
https://render-tron.appspot.com/render/http://my.domain/%3Fpage%3Dhome
```

### Auto detecting loading function

The service detects when a page has loaded by looking at the page load event, ensuring there
are no outstanding network requests and that the page has had ample time to render.

### Rendering budget timeout

There is a hard limit of 10 seconds for rendering. Ensure you don't hit this budget by ensuring
your application is rendered well before the budget expires.

### Web components

Headless Chrome supports web components but shadow DOM is difficult to serialize effectively.
As such, [shady DOM](https://github.com/webcomponents/shadydom) (a lightweight shim for Shadow DOM)
is required for web components.

If you are using web components v0 (deprecated), you will need to enable Shady DOM to
render correctly. In Polymer 1.x, which uses web components v0, Shady DOM is enabled by default.
If you are using Shadow DOM, override this by setting the query parameter `dom=shady` when
directing requests to the Rendertron service.

If you are using web components v1 and either `webcomponents-lite.js` or `webcomponents-loader.js`,
set the query parameter `wc-inject-shadydom=true` when directing requests to the Rendertron
service. This renderer service will force the necessary polyfills to be loaded and enabled.

### Status codes

Status codes from the initial requested URL are preserved. If this is a 200, or 304, you can
set the HTTP status returned by the rendering service by adding a meta tag.

```html
<meta name="render:status_code" content="404" />
```

## Running locally

To install Rendertron and run it locally, first install Rendertron:

```bash
npm install -g rendertron
```

With Chrome installed on your machine run the Rendertron CLI:

```bash
rendertron
```

## Installing & deploying

### Building

Clone and install dependencies:

```bash
git clone https://github.com/GoogleChrome/rendertron.git
cd rendertron
npm install
npm run build
```

### Running locally

With a local instance of Chrome installed, you can start the server locally:

```bash
npm run start
```

### Using the Docker image

After installing docker, build and run the docker image

```bash
docker build -t rendertron . --no-cache=true
docker run -it -p 3000:3000 --name rendertron-container rendertron
```

Load the homepage in any browser:

```bash
http://localhost:3000/
```

Stop the container:

```bash
docker kill rendertron-container
```

Clear containers:

```bash
docker rm -f $(docker ps -a -q)
```

#### Connection error (ECONNREFUSED)

In the case where your kernel lacks user namespace support or are receiving a `ECONNREFUSED` error when trying to access the service in the container (as noted in issues [2](https://github.com/GoogleChrome/rendertron/issues/2) and [3](https://github.com/GoogleChrome/rendertron/issues/3)), the two recommended methods below should solve this:

1. [Recommended] - Use [Jessie Frazelle' seccomp profile](https://github.com/jessfraz/dotfiles/blob/master/etc/docker/seccomp/chrome.json) and `-security-opt` flag
2. Utilize the `--cap-add SYS_ADMIN` flag

[Recommended] Start a container with the built image using Jessie Frazelle' seccomp profile for Chrome:

```bash
wget https://raw.githubusercontent.com/jfrazelle/dotfiles/master/etc/docker/seccomp/chrome.json -O ~/chrome.json
docker run -it -p 3000:3000 --security-opt seccomp=$HOME/chrome.json --name rendertron-container rendertron
```

Start a container with the built image using SYS_ADMIN:

```bash
docker run -it -p 3000:3000 --cap-add SYS_ADMIN --name rendertron-container rendertron
```

To check if your kernel is compatible with Docker, follow [Docker's instructions](https://docs.docker.com/engine/installation/linux/linux-postinstall/#troubleshooting). For CentOS 7, which doesn't have user namespaces enabled, you will [need to enable them](https://github.com/GoogleChrome/rendertron/issues/96#issuecomment-328305721).

#### Chrome crashes due to low shared memory

By default the size of /dev/shm in Docker is 64mb, which may result in Chrome crashes.
To increase the size of /dev/shm you can specify the size to Docker when running the image.

```
docker run ... --shm-size=256m
```

In the future, it will be possible to using /tmp instead of /dev/shm. See the
[Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=736452) for more detail.

### Deploying to Google Cloud Platform

```
gcloud app deploy app.yaml --project <your-project-id>
```

### Deploying using Docker

Rendertron no longer includes a Docker file. Instead, refer to
[Puppeteer documentation](https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker)
on how to deploy run headless Chrome in Docker.

### Config

When deploying the service, set configuration variables by including a `config.json` in the
root. Available configuration options:

-   `datastoreCache` default `false` - set to `true` to enable caching on Google Cloud using datastore
-   `timeout` default `10000` - set the timeout used to render the target page.
-   `port` default `3000` - set the port to use for running and listening the rendertron service. Note if process.env.PORT is set, it will be used instead.
-   `width` default `1000` - set the width (resolution) to be used for rendering the page.
-   `height` default `1000` - set the height (resolution) to be used for rendering the page.

### Troubleshooting

If you're having troubles with getting Headless Chrome to run in your
environment, refer to the
[troubleshooting guide](https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md)
for Puppeteer.


CHANGELOG.md
# Change Log

<!-- ## Unreleased -->

## [2.0.1] 2018-09-18

-   Remove testing and other files from NPM package.
-   Fix NPM main config.
-   Improved restrictions for endpoints.
-   Support for structured data by not stripping all script tags.

## [2.0.0] 2018-07-26

-   Rebuilt with Puppeteer under the hood
-   Rebuilt as Koa server instead of an Express server
-   Rebuilt using Typescript
-   Removed explicit rendering flag
-   Added support for a mobile parameter
-   Added more options for screenshots

## [1.1.1] 2018-01-05

-   Update `debug` flag to log requested URLs to render
-   Fix for renderComplete flag
-   Minor bug fixes

## [1.1.0] 2017-10-27

-   Initial release on NPM


src/test/app-test.ts
/*
 * Copyright 2018 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { test } from 'ava';
import * as Koa from 'koa';
import * as koaStatic from 'koa-static';
import * as path from 'path';
import * as request from 'supertest';

import { Rendertron } from '../rendertron';

const app = new Koa();
app.use(koaStatic(path.resolve(__dirname, '../../test-resources')));

const testBase = 'http://localhost:1234/';

const rendertron = new Rendertron();
let server: request.SuperTest<request.Test>;

test.before(async () => {
    server = request(await rendertron.initialize());
    await app.listen(1234);
});

test('health check responds correctly', async t => {
    const res = await server.get('/_ah/health');
    t.is(res.status, 200);
});

test('renders basic script', async t => {
    const res = await server.get(`/render/${testBase}basic-script.html`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('document-title') !== -1);
    t.is(res.header['x-renderer'], 'rendertron');
});

test('renders script after page load event', async t => {
    const res = await server.get(`/render/${testBase}script-after-load.html`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('injectedElement') !== -1);
});

// This test is failing as the polyfills (shady polyfill & scoping shim) are not
// yet injected properly.
test.failing('renders shadow DOM - no polyfill', async t => {
    const res = await server.get(`/render/${testBase}shadow-dom-no-polyfill.html?wc-inject-shadydom=true`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('shadow-root-text') !== -1);
});

test('renders shadow DOM - polyfill loader', async t => {
    const res = await server.get(`/render/${testBase}shadow-dom-polyfill-loader.html?wc-inject-shadydom=true`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('shadow-root-text') !== -1);
});

test('renders shadow DOM - polyfill loader - different flag', async t => {
    const res = await server.get(`/render/${testBase}shadow-dom-polyfill-loader.html?wc-inject-shadydom`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('shadow-root-text') !== -1);
});

test('renders shadow DOM - webcomponents-lite.js polyfill', async t => {
    const res = await server.get(`/render/${testBase}shadow-dom-polyfill-all.html?wc-inject-shadydom=true`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('shadow-root-text') !== -1);
});

test('script tags and link[rel=import] tags are stripped', async t => {
    const res = await server.get(`/render/${testBase}include-script.html`);
    t.is(res.status, 200);
    t.false(res.text.indexOf('script src') !== -1);
    t.true(res.text.indexOf('injectedElement') !== -1);
    t.false(res.text.indexOf('link rel') !== -1);
    t.true(res.text.indexOf('element-text') !== -1);
});

test('script tags for JSON-LD are not stripped', async t => {
    const res = await server.get(`/render/${testBase}include-json-ld.html`);
    t.is(res.status, 200);
    t.false(res.text.indexOf('script src') !== -1);
    t.true(res.text.indexOf('application/ld+json') !== -1);
    t.false(res.text.indexOf('javascript') !== -1);
});

test('server status code should be forwarded', async t => {
    const res = await server.get('/render/http://httpstat.us/404');
    t.is(res.status, 404);
    t.true(res.text.indexOf('404 Not Found') !== -1);
});

test('http status code should be able to be set via a meta tag', async t => {
    const testFile = 'http-meta-status-code.html';
    const res = await server.get(`/render/${testBase}${testFile}?wc-inject-shadydom=true`);
    t.is(res.status, 400);
});

test('http status codes need to be respected from top to bottom', async t => {
    const testFile = 'http-meta-status-code-multiple.html';
    const res = await server.get(`/render/${testBase}${testFile}?wc-inject-shadydom=true`);
    t.is(res.status, 401);
});

test('screenshot is an image', async t => {
    const res = await server.post(`/screenshot/${testBase}basic-script.html`);
    t.is(res.status, 200);
    t.is(res.header['content-type'], 'image/jpeg');
    t.true(res.body.length > 300);
    t.is(res.body.length, parseInt(res.header['content-length']));
});

test('screenshot accepts options', async t => {
    const res = await server.post(`/screenshot/${testBase}basic-script.html`).send({
        clip: { x: 100, y: 100, width: 100, height: 100 },
        path: 'test.jpeg',
    });
    t.is(res.status, 200);
    t.is(res.header['content-type'], 'image/jpeg');
    t.true(res.body.length > 300);
    t.is(res.body.length, parseInt(res.header['content-length']));
});

test('invalid url fails', async t => {
    const res = await server.get(`/render/abc`);
    t.is(res.status, 403);
});

test('unknown url fails', async t => {
    const res = await server.get(`/render/http://unknown.blah.com`);
    t.is(res.status, 400);
});

test('file url fails', async t => {
    const res = await server.get(`/render/file:///dev/fd/0`);
    t.is(res.status, 403);
});

test('file url fails for screenshot', async t => {
    const res = await server.get(`/screenshot/file:///dev/fd/0`);
    t.is(res.status, 403);
});

test.failing('explicit render event ends early', async t => {
    const res = await server.get(`/render/${testBase}explicit-render-event.html`);
    t.is(res.status, 200);
    t.true(res.text.indexOf('async loaded') !== -1);
});

// TODO: support URL whitelisting.
// test('whitelist ensures other urls do not get rendered', async(t) => {
//   const server = await createServer({
//     renderOnly: [testBase]
//   });
//   let res = await server.get(`/render/${testBase}basic-script.html`);
//   t.is(res.status, 200);

//   res = await server.get(`/render/http://anotherDomain.com`);
//   t.is(res.status, 403);
// });

test('unknown url fails safely on screenshot', async t => {
    const res = await server.get(`/render/http://unknown.blah.com`);
    t.is(res.status, 400);
});


src/test/datastore-cache-test.ts
/*
 * Copyright 2018 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

'use strict';

import { test } from 'ava';
import * as Koa from 'koa';
import * as koaCompress from 'koa-compress';
import * as request from 'supertest';
import * as route from 'koa-route';

import { DatastoreCache } from '../datastore-cache';

const app = new Koa();
const server = request(app.listen());
const cache = new DatastoreCache();

app.use(route.get('/compressed', koaCompress()));

app.use(cache.middleware());

let handlerCalledCount = 0;

test.before(async () => {
    await cache.clearCache();
});

app.use(
    route.get('/', (ctx: Koa.Context) => {
        handlerCalledCount++;
        ctx.body = `Called ${handlerCalledCount} times`;
    }),
);

const promiseTimeout = function(timeout: number) {
    return new Promise(resolve => {
        setTimeout(resolve, timeout);
    });
};

test('caches content and serves same content on cache hit', async t => {
    let res = await server.get('/?basictest');
    const previousCount = handlerCalledCount;
    t.is(res.status, 200);
    t.is(res.text, 'Called ' + previousCount + ' times');

    // Workaround for race condition with writing to datastore.
    await promiseTimeout(500);

    res = await server.get('/?basictest');
    t.is(res.status, 200);
    t.is(res.text, 'Called ' + previousCount + ' times');
    t.truthy(res.header['x-rendertron-cached']);
    t.true(new Date(res.header['x-rendertron-cached']) <= new Date());

    res = await server.get('/?basictest');
    t.is(res.status, 200);
    t.is(res.text, 'Called ' + previousCount + ' times');

    res = await server.get('/?basictest2');
    t.is(res.status, 200);
    t.is(res.text, 'Called ' + (previousCount + 1) + ' times');
});

app.use(
    route.get('/set-header', (ctx: Koa.Context) => {
        ctx.set('my-header', 'header-value');
        ctx.body = 'set-header-payload';
    }),
);

test('caches headers', async t => {
    let res = await server.get('/set-header');
    t.is(res.status, 200);
    t.is(res.header['my-header'], 'header-value');
    t.is(res.text, 'set-header-payload');

    // Workaround for race condition with writing to datastore.
    await promiseTimeout(500);

    res = await server.get('/set-header');
    t.is(res.status, 200);
    t.is(res.header['my-header'], 'header-value');
    t.is(res.text, 'set-header-payload');
});

app.use(
    route.get('/compressed', (ctx: Koa.Context) => {
        ctx.set('Content-Type', 'text/html');
        ctx.body = new Array(1025).join('x');
    }),
);

test('compression preserved', async t => {
    const expectedBody = new Array(1025).join('x');
    let res = await server.get('/compressed').set('Accept-Encoding', 'gzip, deflate, br');
    t.is(res.status, 200);
    t.is(res.header['content-encoding'], 'gzip');
    t.is(res.text, expectedBody);

    // Workaround for race condition with writing to datastore.
    await promiseTimeout(500);

    res = await server.get('/compressed').set('Accept-Encoding', 'gzip, deflate, br');
    t.is(res.status, 200);
    t.is(res.header['content-encoding'], 'gzip');
    t.is(res.text, expectedBody);
});

let statusCallCount = 0;
app.use(
    route.get('/status/:status', (ctx: Koa.Context, status: string) => {
        // Every second call sends a different status.
        if (statusCallCount % 2 === 0) {
            ctx.status = Number(status);
        } else {
            ctx.status = 401;
        }
        statusCallCount++;
    }),
);

test('original status is preserved', async t => {
    let res = await server.get('/status/400');
    t.is(res.status, 400);

    // Non 200 status code should not be cached.
    res = await server.get('/status/400');
    t.is(res.status, 401);
});


middleware/src/test/middleware-test.ts
/*
 * Copyright 2017 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { test } from 'ava';
import * as express from 'express';
import * as net from 'net';
import * as supertest from 'supertest';

import * as rendertron from '../middleware';

/**
 * Start the given Express app on localhost with a random port.
 */
async function listen(app: express.Application): Promise<string> {
    return new Promise<string>((resolve: (url: string) => void) => {
        const server = app.listen(/* random */ 0, 'localhost', () => {
            resolve(`http://localhost:${(server.address() as net.AddressInfo).port}`);
        });
    });
}

/**
 * Make an Express app that uses the Rendertron middleware and returns
 * "fallthrough" if the middleware skipped the request (i.e. called `next`).
 */
function makeApp(options: rendertron.Options) {
    return express()
        .use(rendertron.makeMiddleware(options))
        .use((_req, res) => res.end('fallthrough'));
}

/**
 * Make an Express app that takes the place of a Rendertron server instance and
 * always responds with "proxy <decoded url>".
 */
function makeProxy() {
    return express().use((req, res) => {
        res.end('proxy ' + decodeURIComponent(req.url.substring(1)));
    });
}

const bot = 'slackbot';
const human = 'Chrome';

/**
 * GET a URL with the given user agent.
 */
async function get(userAgent: string, host: string, path: string) {
    return await supertest(host)
        .get(path)
        .set('User-Agent', userAgent);
}

test('makes a middleware function', async t => {
    const m = rendertron.makeMiddleware({ proxyUrl: 'http://example.com' });
    t.truthy(m);
});

test('throws if no proxyUrl given', async t => {
    const makeMiddlewareUntyped = rendertron.makeMiddleware as (options?: unknown) => express.Application;
    t.throws(() => makeMiddlewareUntyped());
    t.throws(() => makeMiddlewareUntyped({}));
    t.throws(() => makeMiddlewareUntyped({ proxyUrl: '' }));
});

test('proxies through given url', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl }));

    const res = await get(bot, appUrl, '/foo');
    t.is(res.status, 200);
    t.is(res.text, 'proxy ' + appUrl + '/foo');
});

test('proxyUrl can have trailing slash', async t => {
    const proxyUrl = await listen(makeProxy());
    // Make sure our other tests are testing the no-trailing-slash case.
    t.false(proxyUrl.endsWith('/'));
    const appUrl = await listen(makeApp({ proxyUrl: proxyUrl + '/' }));

    const res = await get(bot, appUrl, '/foo');
    t.is(res.status, 200);
    t.is(res.text, 'proxy ' + appUrl + '/foo');
});

test('adds shady dom parameter', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl, injectShadyDom: true }));

    const res = await get(bot, appUrl, '/foo');
    t.is(res.status, 200);
    t.is(res.text, 'proxy ' + appUrl + '/foo?wc-inject-shadydom=true');
});

test('excludes static file paths by default', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl }));

    const res = await get(bot, appUrl, '/foo.png');
    t.is(res.text, 'fallthrough');
});

test('url exclusion only matches url path component', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl }));

    const res = await get(bot, appUrl, '/foo.png?params');
    t.is(res.text, 'fallthrough');
});

test('excludes non-bot user agents by default', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl }));

    const res = await get(human, appUrl, '/foo');
    t.is(res.text, 'fallthrough');
});

test('respects custom user agent pattern', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl, userAgentPattern: /borg/ }));

    let res;

    res = await get('humon', appUrl, '/foo');
    t.is(res.text, 'fallthrough');

    res = await get('borg', appUrl, '/foo');
    t.is(res.text, 'proxy ' + appUrl + '/foo');
});

test('respects custom exclude url pattern', async t => {
    const proxyUrl = await listen(makeProxy());
    const appUrl = await listen(makeApp({ proxyUrl, excludeUrlPattern: /foo/ }));

    let res;

    res = await get(bot, appUrl, '/foo');
    t.is(res.text, 'fallthrough');

    res = await get(bot, appUrl, '/bar');
    t.is(res.text, 'proxy ' + appUrl + '/bar');
});

test('forwards proxy error status and body', async t => {
    // This proxy always returns an error.
    const proxyUrl = await listen(express().use((_req, res) => res.status(500).end('proxy error')));
    const appUrl = await listen(makeApp({ proxyUrl }));

    const res = await get(bot, appUrl, '/bar');
    t.is(res.status, 500);
    t.is(res.text, 'proxy error');
});

test('falls through after timeout', async t => {
    // This proxy returns after 20ms, but our timeout is 10ms.
    const proxyUrl = await listen(
        express().use((_req, res) => {
            setTimeout(() => res.end('too slow'), 20);
        }),
    );
    const appUrl = await listen(makeApp({ proxyUrl, timeout: 10 }));

    const res = await get(bot, appUrl, '/foo');
    t.is(res.text, 'fallthrough');
});


src/config.ts
/*
 * Copyright 2018 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License'); you may
 not
 * use this file except in compliance with the License. You may obtain a copy
 of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 under
 * the License.
 */

'use strict';

import * as fse from 'fs-extra';
import * as path from 'path';

const CONFIG_PATH = path.resolve(__dirname, '../config.json');

export type Config = {
    datastoreCache: boolean;
    dynamoCache: boolean;
    timeout: number;
    port: string;
    width: number;
    height: number;
};

export class ConfigManager {
    public static config: Config = {
        datastoreCache: false,
        dynamoCache: true,
        timeout: 10000,
        port: '3000',
        width: 1000,
        height: 1000,
    };

    static async getConfiguration(): Promise<Config> {
        // Load config.json if it exists.
        if (fse.pathExistsSync(CONFIG_PATH)) {
            ConfigManager.config = Object.assign(ConfigManager.config, await fse.readJson(CONFIG_PATH));
        }
        return ConfigManager.config;
    }
}


src/rendertron.ts
import * as Koa from 'koa';
import * as bodyParser from 'koa-bodyparser';
import * as koaCompress from 'koa-compress';
import * as route from 'koa-route';
import * as koaSend from 'koa-send';
import * as koaLogger from 'koa-logger';
import * as path from 'path';
import * as puppeteer from 'puppeteer';
import * as url from 'url';

import { Renderer, ScreenshotError } from './renderer';
import { Config, ConfigManager } from './config';

/**
 * Rendertron rendering service. This runs the server which routes rendering
 * requests through to the renderer.
 */
export class Rendertron {
    app: Koa = new Koa();
    private config: Config = ConfigManager.config;
    private renderer: Renderer | undefined;
    private port = process.env.PORT;

    async initialize() {
        // Load config
        this.config = await ConfigManager.getConfiguration();

        this.port = this.port || this.config.port;

        const browser = await puppeteer.launch({ args: ['--no-sandbox'] });
        this.renderer = new Renderer(browser, this.config);

        this.app.use(koaLogger());

        this.app.use(koaCompress());

        this.app.use(bodyParser());

        this.app.use(
            route.get('/', async (ctx: Koa.Context) => {
                await koaSend(ctx, 'index.html', {
                    root: path.resolve(__dirname, '../src'),
                });
            }),
        );
        this.app.use(route.get('/_ah/health', (ctx: Koa.Context) => (ctx.body = 'OK')));

        // Optionally enable cache for rendering requests.
        if (this.config.datastoreCache) {
            const { DatastoreCache } = await import('./datastore-cache');
            this.app.use(new DatastoreCache().middleware());
        }

        if (this.config.dynamoCache) {
            const { DynamoCache } = await import('./dynamo-cache');
            this.app.use(new DynamoCache().middleware());
        }

        this.app.use(route.get('/render/:url(.*)', this.handleRenderRequest.bind(this)));
        this.app.use(route.get('/screenshot/:url(.*)', this.handleScreenshotRequest.bind(this)));
        this.app.use(route.post('/screenshot/:url(.*)', this.handleScreenshotRequest.bind(this)));

        return this.app.listen(this.port, () => {
            console.log(`Listening on port ${this.port}`);
        });
    }

    /**
     * Checks whether or not the URL is valid. For example, we don't want to allow
     * the requester to read the file system via Chrome.
     */
    restricted(href: string): boolean {
        const parsedUrl = url.parse(href);
        const protocol = parsedUrl.protocol || '';

        if (!protocol.match(/^https?/)) {
            return true;
        }

        return false;
    }

    async handleRenderRequest(ctx: Koa.Context, url: string) {
        if (!this.renderer) {
            throw new Error('No renderer initalized yet.');
        }

        if (this.restricted(url)) {
            ctx.status = 403;
            return;
        }

        const mobileVersion = 'mobile' in ctx.query ? true : false;

        const serialized = await this.renderer.serialize(url, mobileVersion);
        // Mark the response as coming from Rendertron.
        ctx.set('x-renderer', 'rendertron');
        ctx.status = serialized.status;
        ctx.body = serialized.content;
    }

    async handleScreenshotRequest(ctx: Koa.Context, url: string) {
        if (!this.renderer) {
            throw new Error('No renderer initalized yet.');
        }

        if (this.restricted(url)) {
            ctx.status = 403;
            return;
        }

        let options = undefined;
        if (ctx.method === 'POST' && ctx.request.body) {
            options = ctx.request.body;
        }

        const dimensions = {
            width: Number(ctx.query['width']) || this.config.width,
            height: Number(ctx.query['height']) || this.config.height,
        };

        const mobileVersion = 'mobile' in ctx.query ? true : false;

        try {
            const img = await this.renderer.screenshot(url, mobileVersion, dimensions, options);
            ctx.set('Content-Type', 'image/jpeg');
            ctx.set('Content-Length', img.length.toString());
            ctx.body = img;
        } catch (error) {
            const err = error as ScreenshotError;
            ctx.status = err.type === 'Forbidden' ? 403 : 500;
        }
    }
}

async function logUncaughtError(error: Error) {
    console.error('Uncaught exception');
    console.error(error);
    process.exit(1);
}

// Start rendertron if not running inside tests.
if (!module.parent) {
    const rendertron = new Rendertron();
    rendertron.initialize();

    process.on('uncaughtException', logUncaughtError);
}
